### DISCUSSION TITLE
Deploy using Dockerfile

### DISCUSSION THREAD

**Discussion by razukc** [gen-so/Genso.Astrology#7] [Q&A] [answered]
Hello,

Thank you for creating this repository.

I'm trying to deploy it to my hosting that supports Docker. I will post my Dockerfile below. Obviously, it is only a starting point. There may be incomplete codes.

```
# Use the official .NET Core runtime image as the base image
FROM mcr.microsoft.com/dotnet/runtime AS base
WORKDIR /app
# EXPOSE 80

# Copy the published output of the Azure Functions app to the container
FROM mcr.microsoft.com/dotnet/sdk AS build
WORKDIR /src
COPY . /src
RUN dotnet restore
RUN dotnet publish -c Release -o /app/publish

# Use the official Azure Functions runtime image as the base image
FROM mcr.microsoft.com/azure-functions/dotnet AS final
ENV AzureWebJobsScriptRoot=/app/publish
ENV AZURE_FUNCTIONS_ENVIRONMENT Production
ENV ASPNETCORE_URLS=http://*:80/
COPY --from=build /app/publish /app/publish
# Set the default command to run the Azure Functions app
# CMD ["dotnet", "API.dll"]
```

At this point, the deploy is successful but the GUI shows following error:

`Function host is not running.`

Lets try to make it an working example. Please, help!

**Comment by sengiv** [comment, marked as answer]
Thank you for using the repo. 🙏

This is done. You can now run the API in Docker,
check out the latest commit a7869b1557e75fc98c245e72be316d981bfcb16a
View [demo here](https://github.com/VedAstro/VedAstro/discussions/8)

**Reply by razukc** [reply]
Yes, I can confirm that it is working. Good Job! Thank you!

**Reply by sengiv** [reply]
You're welcome  😃 

Btw, we're looking for a 2nd live API server (to separate "Stable" from "Beta" release),
if you have a docker container to spare, please let us know. It would help the project,
increasing stability of the website during updates.

**Reply by razukc** [reply]
One way I can propose is a Web Service on Render.com. If you're familiar with Render, you already know what I'm talking about. It is also what I'm using now to test this API service. So, in summary, you'll need to provide my Github account an access to your repo. That way I can add the repo you shared to Render so that it will auto deploy on changes to your repo. I suggest you make a separate branch for "Beta" release. We can tweak this idea for that you're comfortable with. Let me know. Thanks again.

**Reply by sengiv** [reply]
thank you very much, this will definitely help

> you'll need to provide my Github account an access to your repo

invited, please check

> separate branch for "Beta" release.

yes, will do, right now master will be stable release

**Reply by razukc** [reply]
Thank you for the invitation. accepted and also requested to "Install Render" on Github. You'll have to accept it.

**Reply by sengiv** [reply]
> and also requested to "Install Render" on Github. You'll have to accept it.

alright, done

> If you're familiar with Render, you already know what I'm talking about.

actually I've never used Render before, so you'll have to guide me on this

1. now that the repo is accepted does it mean all future commits will get
auto compiled and published to Render's docker server?

2. how do i view the server URL's, so that I can point vedastro.org to it

can reach me at contact@vedastro.org

**Comment by razukc** [comment]
> 1. now that the repo is accepted does it mean all future commits will get auto compiled and published to Render's docker server?

Yes, that is correct. All future commits will get compiled and published to Render's docker server, like you mentioned. We can choose what branch though. Thus, mentioned "Beta" branch in earlier conversation.

One more thing, you'll have to install Render to your Github yourself as only owner are allowed to do so. Did not know about this earlier. Here is the [link](https://github.com/apps/render/installations/new)
> 2. how do i view the server URL's, so that I can point vedastro.org to it

I can add a custom domain in Render to achieve this. You'll have to supply a domain name and make necessary DNS configuration at your end. Here is the [wiki](https://render.com/docs/custom-domains#configuring-dns-to-point-to-render)


Sorry, I'll have to setup Whatsapp on my PC. Haven't done that. I will try to set that up for future conversations. I will reply to your email next time.

Thanks.

**Reply by sengiv** [reply]
>  you'll have to install Render to your Github yourself as only owner are allowed to do so.

it says already installed 44 min ago, btw made you owner, maybe you can try again

![image](https://user-images.githubusercontent.com/43817262/222957998-5e923372-c78f-4076-9295-e9cc7adbb278.png)

**Reply by razukc** [reply]
Congratulations! Your repo has been successfully deployed to [Render](https://vedastro-api.onrender.com). 
Now, If I may suggest, you can change the domain name to something like api.vedastro.org. Let me know when you're ready. Thank you so much again.

**Reply by sengiv** [reply]
> Congratulations! Your repo has been successfully deployed to [Render](https://vedastro-api.onrender.com/).

Much appreciated 🙏
What about the price? I understand that Render is not free, so if you would like we can try to pay
you back when there's a donation received.

</br>

> something like api.vedastro.org. Let me know when you're ready.

There is an Azure CDN in place that redirects traffic,
so I'll just add `vedastro-api.onrender.com` to the rule engine.
It should be accessible soon via `vedastro.org/api/...`

</br>

> compiled and published to Render's docker server, like you mentioned. We can choose what branch though. Thus, mentioned "Beta" branch in earlier conversation.

Sorry to bother you, but after more than 5 hours, the latest commits from `master` or `beta` is not being deployed.
Maybe, branch name wrongly selected in Render? 🤔 

![image](https://user-images.githubusercontent.com/43817262/222980997-e9c43b7c-a816-4ba0-91d2-ab85cc94dc76.png)

**Reply by razukc** [reply]
Render log shows that there have been some deployments. Some exited with error. However, the latest deployment was successful. There is this thing with Render that it will fail deployment if there is some error in the software. I'll find some way to share the logs with you. 

Currently, `master` branch is configured to accept updates from Git. Let me know if you want to change it to `beta` at any point.

About Render Pricing, its a basic Starter Plan. I can take care of this for a few months.

**Comment by sengiv** [comment]
> However, the latest deployment was successful.

Nice ✈️  thanks for the update

> error in the software. I'll find some way to share the logs with you.

**It's not very important**, but if you want to experiment :- link a Slack Webhook to Docker.
Automatically any logs from Docker will appear as messages in a Slack channel.
You could have 1 channel for 1 project, so monitor all builds from many places not just in Render,
but also from other docker builds elsewhere. If you're interested there's code in API doing
this for certificate management.

> Currently, master branch is configured to accept updates from Git.
Let me know if you want to change it to beta at any point.

Perfect, `master` is the correct branch for stable commits

> About Render Pricing, its a basic Starter Plan. I can take care of this for a few months.

Even if you took care of the bill for a day, it would mean a lot to us. 🙏 So thanks again.
Btw for your info, your Render server endpoint is now live routed via a _**load balancer**_ behind `vedastro.org/api`.
This means only low latency requests from near your Render server location, will be directed to this Docker instance.
If everything works as expected, the estimated cost for running this Docker image **24/7** will be close to only **$2/year** 💰.
Which is good news 😄

**Comment by kingcarol1** [comment]
@razukc any idea here? https://github.com/VedAstro/VedAstro/issues/186 [tldr: i am trying to deploy the api but idk how, in the end i want to make requests through python to the api; i tried docker image on railway and importing from github repo; did not work; docker image kinda works but its limited it doesnt support python calculate methods]
